core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java

```diff
@@ -85,28 +72,40 @@ public final void addListener(Consumer<? super T> listener) {
      */
     public final void addListener(Consumer<? super T> listener, boolean notifyLatestValue) {
         requireNonNull(listener, "listener");
-        reentrantLock.lock();
-        try {
-            if (notifyLatestValue) {
-                final T latest = latestValue();
-                if (latest != null) {
-                    listener.accept(latest);
-                }
+        if (notifyLatestValue) {
```

Comment:
Question) I'm wondering if CAS is really safe in this case since we may call `removeListener` from listeners like done in `AbstractEndpointSelector`

https://github.com/line/armeria/blob/9ddd6c1b76fdc0fe8943dd417e80773cb92eebdc/core/src/main/java/com/linecorp/armeria/client/endpoint/AbstractEndpointSelector.java#L141

```java
static class MyListener implements Consumer<List<Endpoint>> {

    final EndpointGroup group;
    int invoked = 0;
    MyListener(EndpointGroup group) {
        this.group = group;
    }

    @Override
    public void accept(List<Endpoint> endpoints) {
        invoked++;
        group.removeListener(this);
    }

    public int invoked() {
        return invoked;
    }
}

@Test
void shouldNotifyListenerWithLatestValue2() throws InterruptedException {
    final DynamicEndpointGroup group = new DynamicEndpointGroup();
    group.setEndpoints(ImmutableList.of(Endpoint.of("foo")));
    final MyListener listener = new MyListener(group);
    group.addListener(listener, true);
    assertThat(listener.invoked).isEqualTo(1);
    // listener isn't removed at this point, possible leak
}
```

Reply:
Good point. Would it be safe after the CAS operation for adding is done?
https://github.com/line/armeria/blob/a2be9d95e16319a3d4ebe5928bdf9dd6c4fa7422/core/src/main/java/com/linecorp/armeria/common/util/AbstractListenable.java#L85-L85

Reply:
I guess so.

I'm also a little wary that the count of `listeners` can be high and cause some contention (since `listeners` is bound by the number of requests which can be large for all we know)

Was this change due to a bug? or was it a refactoring?

Reply:
> since listeners is bound by the number of requests which can be large for all we know

In normal cases, the number of listeners is not bound by requests. `ListeningFuture` will only be used when an `EndpointGroup` is not initialized or becomes empty because of an outage. 

It was changed to avoid invoking external callbacks with the lock which could cause potential deadlock.

Reply:
After looking at AbstractEndpointSelector, I don't think we need to add a new listener for every `select()` method for notification. Let me tweak it a bit.

Reply:
I see, we're not doing any scheduling to other threads inside the lock so I didn't think there was a possibility of deadlock.
I guess it's fine to leave as is then.

Reply:
Fixed. (Sorry for the frequent changes. 😅)